src/Foundatio/Messaging/MessageBusBase.cs

```diff
@@ -254,8 +269,14 @@ protected class Subscriber {
             public Type Type { get; set; }
             public Func<object, CancellationToken, Task> Action { get; set; }
 
-            public bool IsAssignableFrom(Type type) {
-                return _assignableTypesCache.GetOrAdd(type, t => Type.GetTypeInfo().IsAssignableFrom(t));
+            public Func<object, bool> MessageFilter { get; set; }
+
+            public bool IsAssignableFrom(Type type, object message) {
+
+                if (MessageFilter == null)
```

Comment:
I guess it's more readable to set this, but I think perf wise when this is called a billion times it would be better to just null check it. Also is there any concern that code in the message filter might throw? I think we should make the assumption that it will never throw and don't put stupid code in there. I also don't think it needs to be async but I can't think of any use case where it would need to be.

Reply:
```suggestion
                  if (MessageFilter == null)
                    return _assignableTypesCache.GetOrAdd(type, t => Type.GetTypeInfo().IsAssignableFrom(t));

                return _assignableTypesCache.GetOrAdd(type, t => Type.GetTypeInfo().IsAssignableFrom(t) && MessageFilter(message));
```
The MessageFilter is not async. It is a Function

Reply:
Yeah, I know that. I'm just thinking out loud if there is ever a valid use case where a filter would need to be async.

Reply:
:-)
you should better do the async action in the message handler. not in the filter